[UniformGridLayout] Fix bug where controls would not know their limitations when being evaluated for layouting

chingucoding opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
The issue was that we don't actually limit a control's size, instead we ask them how large they want to be and use that. When the user sets the MinItemHeight or MinItemWidth, this is not being respected and images would think they have all the space they want when in reality, they do not. This PR fixes that by setting MaxWidth and MaxHeight beforehand.
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
<!--- Use the syntax "Closes #1234" or "Fixes #5678" so that GitHub will close the issue once the PR is complete. -->
Closes #5154 
## How Has This Been Tested?
<!--- Please describe how you tested your changes. -->
Tested manually (test code checked in).
## Screenshots (if appropriate):
<!--- If you are making visual changes to a Control or adding a new Control, please include screenshots showing the result. -->
## Additional notes
Why are the properties MinItemHeight and MinItemWidth called Min... if in reality items also are not allowed to exceed that width?

Edit: I was forgetting about a certain thing called item stretch...

beervoley commented:
/azp run

Ryken100 commented on dev/Repeater/UniformGridLayoutState.cpp:
```diff
@@ -42,24 +42,46 @@ void UniformGridLayoutState::EnsureElementSize(
+        }
+        else {
+            if (!isnan(itemHeight)) {
+                frElement.MaxWidth(itemHeight);
```
Should `MaxHeight` be set here instead?

Ryken100 commented on dev/Repeater/UniformGridLayoutState.cpp:
```diff
@@ -42,24 +42,46 @@ void UniformGridLayoutState::EnsureElementSize(
+    if (const auto frElement = element.try_as<winrt::FrameworkElement>()) {
+        if (orientation == winrt::Orientation::Horizontal) {
+            if (!isnan(itemWidth)) {
+                frElement.MaxWidth(itemWidth);
```
Please correct me if I'm wrong, but is this essentially setting `MaxWidth` to `UniformGridLayout.MinItemWidth`? If that's the case, I think it would be best to borrow code from `SetSize()` below to get the effective size based on `UniformGridLayout.ItemsStretch`, as the effective size won't always match `MinItemWidth` and `MinItemHeight`.

Additionally, I wonder if setting `MaxWidth` and `MaxHeight` would interfere with cases where the user has already set those values in their `ItemTemplate`. Perhaps instead of setting `MaxWidth` and `MaxHeight`, this method could return a `Size` which would be used to measure `realizedElement` in `EnsureElementSize()`.

chingucoding replied:
Good idea, changed the code to the corresponding `winrt::Size` object to the elements `Measure`.

Ryken100 replied:
I see, that's a good change, but unfortunately I'm not sure if it will fix the issue when `UniformGridLayout.ItemsStretch = Fill`, as the width of a column won't be equal to `UniformGridLayout.MinItemWidth`. We need to calculate the actual width of a column, and then use that value to measure the element.

For example, if `UniformGridLayout.MinItemWidth = 250`, and the `ItemsRepeater` is in a 1200px width container, I think the logic to determine the column width would go something like this:
* Divide container width by `MinItemWidth` to get column count: 1200 / 250 = 4
* Divide container width by column count to get effect width: 1200 / 4 = 300
* Measure first element using `new Size(300, availableWidth.Height)`

You would have columns that are 300px wide, not 250px wide, so measuring them with `MinItemWidth` alone would create incorrect results. `MinColumnSpacing` and `MinRowSpacing` would have to be accounted for as well.

I'm happy help with this if you'd like :)

chingucoding replied:
Very valid point. I updated the code now so that we take stretching into account. Please let me know your thoughts @Ryken100! Thank you for helping me with this PR :)

ranjeshj commented:
/azp run

ranjeshj commented:
@chingucoding It has Min in the name because if there is extra space remaining in the line, then depending on the value for stretch, each of the items could get more space.

chingucoding commented:
> @chingucoding It has Min in the name because if there is extra space remaining in the line, then depending on the value for stretch, each of the items could get more space.

Fair enough, forgot about the stretch modes that also allow more space for the elements.

JustinXinLiu commented:
@StephenLPeters @jevansaks @ranjeshj can you please help review this?

ranjeshj commented on dev/Repeater/UniformGridLayoutState.cpp:
```diff
@@ -42,28 +42,71 @@ void UniformGridLayoutState::EnsureElementSize(
     {
         // If the first element is realized we don't need to get it from the context
-        if (auto realizedElement = m_flowAlgorithm.GetElementIfRealized(0))
+        if (auto realizedElement = m_flowAlgorithm.GetElementIfRealized(0).as<winrt::FrameworkElement>())
```
Is this cast required ?

ranjeshj commented on dev/Repeater/UniformGridLayoutState.cpp:
```diff
@@ -42,28 +42,71 @@ void UniformGridLayoutState::EnsureElementSize(
         {
             // Not realized by flowlayout, so do this now!
-            if (const auto firstElement = context.GetOrCreateElementAt(0, winrt::ElementRealizationOptions::ForceCreate))
+            if (const auto firstElement = context.GetOrCreateElementAt(0, winrt::ElementRealizationOptions::ForceCreate).as<winrt::FrameworkElement>())
```
Is this cast required ?

ranjeshj commented on dev/Repeater/UniformGridLayoutState.cpp:
```diff
@@ -42,28 +42,71 @@ void UniformGridLayoutState::EnsureElementSize(
+    double minRowSpacing,
+    double minColumnSpacing)
+{
+    if (orientation == winrt::Orientation::Horizontal) {
```
nit: newline for block here and below to keep the formatting consistent across the file.

ranjeshj replied:
can you add a small comment explaining why we are doing this when trying to calculate the available size/ensure size.

ranjeshj commented on dev/Repeater/UniformGridLayoutState.cpp:
```diff
@@ -42,28 +42,71 @@ void UniformGridLayoutState::EnsureElementSize(
+    return availableSize;
+}
+
+double UniformGridLayoutState::CalculateExtraPixelsPerColumn(unsigned int maxItemsPerLine,
```
nit: CalculateExtraPixelsInLine ? since it is row or column based on orientation.

chingucoding commented:
Addressed your feedback @ranjeshj. Regarding the test, I'm stuck right now. Creating an image template from code behind doesn't work as the images don't seem to request more than a size of 0px*0px. I was trying this:

```C#
ItemsRepeater repeater = null;
            ScrollViewer scroller = null;
            RunOnUIThread.Execute(() =>
            {
                repeater = new ItemsRepeater() {
                    ItemsSource = Enumerable.Range(0,6), // no items 
                    Layout = new UniformGridLayout() { 
                        Orientation = Orientation.Vertical,
                        MaximumRowsOrColumns = 1,
                        MinItemWidth = 100
                    },
                    MaxWidth=300,
                    ItemTemplate = GetDataTemplate("<Image Source=\"/Images/recipe1.png\" MinWidth=\"100\"/>")
                };

                scroller = new ScrollViewer() {
                    Content = repeater
                };

                Content = scroller;
                repeater.UpdateLayout();
                Content.UpdateLayout();
            });

            IdleSynchronizer.Wait();

            RunOnUIThread.Execute(() =>
            {
                // Ensure we do not crash and get zero size.
                // Assert not implemented yet
            });
```

Maybe you guys know more? The problem is that I need a control that actually needs to depend on the width being smaller and as such using less space (i.e. aspect ratio), that's  why I was trying to use an image here.

StephenLPeters commented:
@chingucoding what behavior do you need from image? I don't understand the description of the requirement.

chingucoding commented:
@StephenLPeters I need a control that when given less width also renders with less height. A control that has a fixed aspect ratio is meeting that requirement which is way I was using an Image control for testing.

StephenLPeters commented:
AnimatedVisualPlayer has the same stretch property that image does and AnimatedIcon is hard coded to do a Uniform scale mode.  @codendone is @chingucoding missing something for getting images from code behind to render properly?

ranjeshj commented:
> AnimatedVisualPlayer has the same stretch property that image does and AnimatedIcon is hard coded to do a Uniform scale mode. @codendone is @chingucoding missing something for getting images from code behind to render properly?

@chingucoding  You can also create a custom panel that behaves similar to image to simulate this.

chingucoding commented:
I was trying to avoid having to write a custom control for this, but guess there is no other way. Unfortunately AnimatedVisualPlayer also seems to request some odd size. With a custom presenter though, the ItemsRepeater still ends up with a desired size of 0 and an actual height of 664, regardless of the content. I'm quite confused why the desiredsize remains zero while the actual height stayed at 664 the whole time. Maybe I am missing something crucial here? Test is also pushed to this PR now.

StephenLPeters commented:
@chingucoding can you merge main to pick up a infrastructure fix?

chingucoding commented:
@StephenLPeters Done

StephenLPeters commented:
/azp run

ranjeshj commented on dev/Repeater/APITests/FlowLayoutTests.cs:
```diff
@@ -1260,6 +1259,43 @@ public void ValidateUniformGridWithNoItems()
+
+                Content = scroller;
+                repeater.UpdateLayout();
+                Content.UpdateLayout();
```
should not be required, calling updatelayout on any element runs an entire layout.

ranjeshj commented on dev/Repeater/APITests/FlowLayoutTests.cs:
```diff
@@ -1260,6 +1259,43 @@ public void ValidateUniformGridWithNoItems()
+            {
+                // Assert not implemented yet
+                Verify.IsTrue(Math.Abs(300 - repeater.DesiredSize.Width) < 1, "Repeater width should be 300");
+                Verify.IsTrue(repeater.DesiredSize.Height > 1, "Repeater should have height");
```
looks like this condition failed in the test run.

StephenLPeters commented:
I've attempted to fix this test but am having huge ammounts of issues getting the test to run locally.  I can't get XamlReader.Load to recognize that AspectRatioRespectingControl is a type.  I think this is a symptom of an issue in our build system that causes tests run with the TestExplorer to not update when you build the project.  I was able to get a type to be recognized once today, but for the life of me I don't know how I did it or what to do to reproduce the success...

StephenLPeters commented:
/azp run

StephenLPeters commented:
/azp run

StephenLPeters commented:
@chingucoding any idea whats going on with this test failure?  This isn't the same failure that we saw before my changes and I don't understand how my change could have caused this... have you seen it locally?

chingucoding commented:
I have not seen these error messages before at all. Completely lost as to why your code changes would cause these errors. My guess would be rerun the tests and see if they occur again? But even if it works then, could it be that we are introducing an unreliable test?

StephenLPeters commented:
I'm not sure, I debugged this for hours already and am totally lost.  :(

ranjeshj commented:
> AspectRatioRespectingControl

This might be a case where the IXMP information for this type is not generated because it is not used from markup. I think adding the Bindable attribute on the type will force the metadata to be generated. @RealTommyKlein  ?

RealTommyKlein commented:
> > AspectRatioRespectingControl
> 
> This might be a case where the IXMP information for this type is not generated because it is not used from markup. I think adding the Bindable attribute on the type will force the metadata to be generated. @RealTommyKlein ?

Yes, that sounds correct.  If there isn't any `.xaml` file that uses `AspectRatioRespectingControl` but it still needs to be created from `XamlReader.Load`, `AspectRatioRespectingControl` needs the `Windows.UI.Xaml.Data.Bindable` attribute added to it so the Xaml compiler is forced to generate type information for it so the framework can use it.

chingucoding commented:
That's very interesting, thank you for your help @RealTommyKlein !

I've added the attribute now, can we rerun? (Strangely, on my machine, the test works fine)

StephenLPeters commented:
/azp run

StephenLPeters commented:
@RealTommyKlein This didn't fix the issue for me locally

RealTommyKlein commented:
> @RealTommyKlein This didn't fix the issue for me locally

You can sanity-check that information for the type is generated in `XamlTypeInfo.g.cs`, but otherwise sounds like there's another issue. You could try simplifying the markup passed into `XamlReader.Load` to isolate what's causing the problem.

StephenLPeters commented:
> > @RealTommyKlein This didn't fix the issue for me locally
> 
> You can sanity-check that information for the type is generated in `XamlTypeInfo.g.cs`, but otherwise sounds like there's another issue. You could try simplifying the markup passed into `XamlReader.Load` to isolate what's causing the problem.

I don't see the type in XamlTypeInfo.g.cs either.

chingucoding commented:
> > > @RealTommyKlein This didn't fix the issue for me locally
> > 
> > 
> > You can sanity-check that information for the type is generated in `XamlTypeInfo.g.cs`, but otherwise sounds like there's another issue. You could try simplifying the markup passed into `XamlReader.Load` to isolate what's causing the problem.
> 
> I don't see the type in XamlTypeInfo.g.cs either.

Could it be that the type also has to be made public for this to work?

StephenLPeters commented:
That does cause the type info to be present but doesn't resolve the test issue for me

chingucoding commented:
And what if we simply add the control to one of the XAML pages so that it DEFINITELY is exposed to the XAML loader/compiler?

StephenLPeters commented:
I tried that already, it compiles but crashes when you navigate to the page.

chingucoding commented:
And with which exception does it crash? Is it something in the area of not able to load it or could it be that we are actually messing with some rendering stuff that crashes the app in that case?

StephenLPeters commented:
Its a xaml parse exception.

chingucoding commented:
Hmm that's really strange... I'll try and see if I can repro it. Could it be due to different VS or OS versions? After all, the test only failed on some of the test runners and not on all if I recall correctly.

RealTommyKlein commented:
@StephenLPeters pointed me in the right direction - this is caused by the control being in the `Windows.UI.Xaml` namespace, so the Xaml compiler believes it's built in to the framework and the framework doesn't need generated type information to use it.  If the control is moved into a different namespace like `Tests.MUXControls.ApiTests.RepeaterTests.Common` instead of `Windows.UI.Xaml.Tests.MUXControls.ApiTests.RepeaterTests.Common`, it should work.

chingucoding commented:
I've moved the type to a different namespace now and adjusted the test accordingly. (Tests still runs but that's not too surprising)

StephenLPeters commented:
/azp run
